src/hierarchical.rs

```diff
@@ -0,0 +1,493 @@
+                    a = i;
+                }
+            } else {
+                em[i] = INFINITY;
+            }
+        }
+
+        let mut b = replace(&mut pi[a], n);
+        let mut c = replace(&mut lambda[a], em[a]);
+
+        if a < n - 1 {
+            while  b < n - 1 {
+                b = replace(&mut pi[b], n);
+                c = replace(&mut lambda[b], c);
+            }
+
+            if b == n - 1 {
+                pi[b] = n;
+                lambda[b] = c;
+            }
+        }
+
+        for i in (0..n) {
+            if pi[pi[i]] == n && lambda[i] >= lambda[pi[i]] {
+                pi[i] = n;
+            }
+        }
+    })
+}
+
+
+fn using_pointer_representation<'a, P, F>(points: &Vec<&'a P>, f: F) -> Box<Dendrogram<&'a P>>
+    where P: Point, F: Fn(usize, &mut Vec<usize>, &mut Vec<f64>, &mut Vec<f64>)
+{
+    let mut pi = vec![0; points.len()];
+    let mut lambda = vec![0.0; points.len()];
+    let mut em = vec![0.0; points.len()];
+
+    pi[0] = 0;
+    lambda[0] = INFINITY;
+
+    for n in (1..points.len()) {
+        pi[n] = n;
+        lambda[n] = INFINITY;
+
+        for i in (0..n) {
+            em[i] = points[i].dist(points[n]);
+        }
+
+        f(n, &mut pi, &mut lambda, &mut em);
+    }
+
+    // convert pointer representation to dendrogram
+    let mut dendrograms: Vec<_> = (0..points.len()).map(|i|
+        Some(Box::new(Dendrogram::Leaf(points[i])))).collect();
+
+    let mut idx: Vec<_> = (0..points.len()).collect();
+    idx.sort_by(|&a, &b| lambda[a].partial_cmp(&lambda[b]).unwrap());
+
+    for i in (0..points.len()-1) {
+        let leaf = idx[i];
+        let merge = pi[leaf];
+
+        let a = replace(&mut dendrograms[leaf], None).unwrap();
+        let b = replace(&mut dendrograms[merge], None).unwrap();
+
+        if leaf < merge {
+            dendrograms[merge] = Some(Box::new(Dendrogram::Branch(lambda[leaf], a, b)));
+        } else {
+            dendrograms[leaf] = Some(Box::new(Dendrogram::Branch(lambda[leaf], b, a)));
+        }
+    }
+
+    dendrograms.swap_remove(points.len() - 1).unwrap()
+}
+
+
+#[cfg(test)]
+#[allow(non_snake_case)]
+fn minimal_distance(d: &Fn(usize, usize) -> f64, A: &[usize], B: &[usize]) -> f64 {
+    A.iter().flat_map(|&a|
+        B.iter().map(move |&b| Dist(d(a,b)))
+    ).min().unwrap().0
+}
+
+#[allow(non_snake_case)]
+fn maximal_distance(d: &Fn(usize, usize) -> f64, A: &[usize], B: &[usize]) -> f64 {
+    A.iter().flat_map(|&a|
+        B.iter().map(move |&b| Dist(d(a,b)))
+    ).max().unwrap().0
+}
+
+#[derive(PartialEq, PartialOrd)]
+struct Dist(f64);
+impl Eq for Dist { }
+impl Ord for Dist {
+    fn cmp(&self, other: &Self) -> Ordering {
+        self.partial_cmp(other).unwrap()
+    }
+}
+
+
+/// A hierarchical tree structure describing the merge operations.
+#[derive(Clone, Debug)]
+pub enum Dendrogram<T> {
+    /// The union of two sub-trees (clusters) and their distance.
+    Branch(f64, Box<Dendrogram<T>>, Box<Dendrogram<T>>),
```

Comment:
Could you elaborate how you got the 15%? It seems pretty surprising that so much time is spent in the allocation when the real work is happening in the linkage. Considering that there should only be `2n` allocations.

I do like the initial representation, because it states the Dendrogram very clear. One doesn't have to reason about potential indices or whatnot, because it's all in the types. Furthermore, CLINK/SLINK use their pointer representation during clustering - which works with a lot of indices - and use the Dendrogram representation only during post-processing for a better interface. It shouldn't matter for the naive implementation that much either, because (as the name suggests) there are parts with a much higher complexity that should out-weight the allocation penalties by a lot.

So unless it isn't a real performance concern, I would refrain from changing the actual representation. Again the question, how did you arrive at the number of 15% ?